### PR TITLE
Feature/634 casl prisma authz

### PR DESCRIPTION
**Pull request by skosito**
# Description

This PR further extends on usage of casl library for authz.
- added casl prisma package to be able to define conditions on prisma models (https://casl.js.org/v6/en/package/casl-prisma) and modified authz factory accordingly
- made authz guard more generic with generic prisma query to fetch subject that is being affected 
- fixed the bug that caused public endpoints to still have this guard applied on

It contains these authz rules for controllers (assuming that internal bpi subject can manage all, this is just for external):
bpi subject - itself
workgroups - bpi subject that is workgroup admin
workflow/workstep - bpi subject that is workgroup admin that these are associated with
messages/transactions - probably best to leave to internal subject for now
subject account - owner bpi subject
bpi account - owner of associated bpi subject accounts owner

Purpose is to have minimal rules set for milestone 3, and if this is ok approach we could extend if needed when we have use case for milestone 4 with same/similar approach for new rules.

## Related Issue

https://github.com/eea-oasis/baseline/issues/634
https://github.com/eea-oasis/baseline/issues/635
https://github.com/eea-oasis/baseline/issues/636

## Motivation and Context

Check issues

## How Has This Been Tested

Using postman and various scenarios:
- internal bpi subject can manage all
- external can only crud its own bpi subject entity and workgroups where he is admin (this is my assumption)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Request to be added as a Code Owner/Maintainer

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I commit to abide by the Responsibilities of Code Owners/Maintainers.

### CODE REVIEW

File: examples/bri-3/src/bri/workgroup/workflows/api/workflows.controller.ts [hunk 1]
```diff
@@ -22,16 +23,19 @@ export class WorkflowController {
...
   }
 
   @Get()
+  @CheckAuthz({ action: 'read', type: 'Workflow' })
   async getAllWorkflows(): Promise<WorkflowDto[]> {
     return await this.queryBus.execute(new GetAllWorkflowsQuery());
   }
 
   @Post()
+  @CheckAuthz({ action: 'manage', type: 'Workflow' })
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/workgroup/workflows/api/workflows.controller.ts:38]: why not create?

File: examples/bri-3/src/bri/workgroup/workgroups/api/workgroups.controller.ts [hunk 2]
```diff
@@ -17,17 +17,20 @@ import { ArchiveWorkgroupCommand } from '../capabilities/archiveWorkgroup/archiv
...
   constructor(private commandBus: CommandBus, private queryBus: QueryBus) {}
 
   @Get('/:id')
+  @CheckAuthz({ action: 'read', type: 'Workgroup' })
   async getWorkgroupById(@Param('id') id: string): Promise<WorkgroupDto> {
     return await this.queryBus.execute(new GetWorkgroupByIdQuery(id));
   }
 
   @Post()
+  @CheckAuthz({ action: 'manage', type: 'Workgroup' })
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/workgroup/workgroups/api/workgroups.controller.ts:33]: Why not create?

File: examples/bri-3/src/bri/workgroup/worksteps/api/worksteps.controller.ts [hunk 3]
```diff
@@ -25,16 +26,19 @@ export class WorkstepController {
...
   }
 
   @Get()
+  @CheckAuthz({ action: 'read', type: 'Workstep' })
   async getAllWorksteps(): Promise<WorkstepDto[]> {
     return await this.queryBus.execute(new GetAllWorkstepsQuery());
   }
 
   @Post()
+  @CheckAuthz({ action: 'manage', type: 'Workstep' })
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/workgroup/worksteps/api/worksteps.controller.ts:41]: Why not create?

File: examples/bri-3/src/bri/identity/bpiSubjects/api/subjects.controller.ts [hunk 4]
```diff
@@ -23,17 +23,19 @@ export class SubjectController {
...
 
   @Get()
+  @CheckAuthz({ action: 'read', type: 'BpiSubject' })
   async getAllBpiSubjects(): Promise<BpiSubjectDto[]> {
     return await this.queryBus.execute(new GetAllBpiSubjectsQuery());
   }
 
   @Post()
-  @CheckAuthz({ action: 'manage', subject: 'BpiSubject' })
+  @CheckAuthz({ action: 'create', type: 'BpiSubject' })
```
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/identity/bpiSubjects/api/subjects.controller.ts:38]: manage here if create is not defined as an action?

**Review by ognjenkurtic — APPROVED** [review]

File: examples/bri-3/src/bri/workgroup/workgroups/api/workgroups.controller.ts [hunk 5]
```diff
@@ -62,11 +66,13 @@ export class WorkgroupController {
   }
 
   @Put('archive/:id')
+  @CheckAuthz({ action: 'delete', type: 'Workgroup' })
```
[the remark this answers is not in the document]
Comment (skosito) [reply examples/bri-3/src/bri/workgroup/workgroups/api/workgroups.controller.ts:69]: i also dont think we need separate

File: examples/bri-3/src/bri/authz/authz.factory.ts [hunk 6]
```diff
@@ -1,42 +1,78 @@
...
-  typeof actions[number],
-  (
-    | typeof subjects[number]
-    | ForcedSubject<Exclude<typeof subjects[number], 'all'>>
-  ),
-];
-export type AppAbility = PureAbility<AppAbilities>;
+type AppSubjects =
+  | 'all'
+  | Subjects<{
```
[the remark this answers is not in the document]
Comment (skosito) [reply examples/bri-3/src/bri/authz/authz.factory.ts:17]: i will add them here, but dont have specific rules, they will automatically be only allowed to be managed by internal bpi subject, we can make it more granular after if needed

**Review by ognjenkurtic — APPROVED** [review]

File: examples/bri-3/src/bri/workgroup/workgroups/api/workgroups.controller.ts (hunk 5 above)
Comment (Kasshern) [inline examples/bri-3/src/bri/workgroup/workgroups/api/workgroups.controller.ts:69]: Why is this delete?

**Review by Kasshern — APPROVED** [review]
Beautiful. How much manual testing has gone into making sure this works as intended?

**Pull request merged**